File: java/twelve-days/src/main/java/TwelveDays.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class TwelveDays {
    private final String verseStart = "On the %s day of Christmas my true love gave to me: ";
    private final String[] verseNumbers = { "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth" };
    private final String[] verseEnds = {
            "twelve Drummers Drumming, ",
            "eleven Pipers Piping, ",
            "ten Lords-a-Leaping, ",
            "nine Ladies Dancing, ",
            "eight Maids-a-Milking, ",
            "seven Swans-a-Swimming, ",
            "six Geese-a-Laying, ",
            "five Gold Rings, ",
            "four Calling Birds, ",
            "three French Hens, ",
            "two Turtle Doves, and ",
            "a Partridge in a Pear Tree."
    };

    String verse(int verseNumber) {
        String[] verses = Arrays.copyOfRange(
                verseEnds,
                verseEnds.length - verseNumber,
                verseEnds.length
        );

        String end = String.join("", verses) + "\n";
        return String.format(verseStart, verseNumbers[verseNumber - 1]) + end;
    }

    String verses(int startVerse, int endVerse) {
        List<String> output = new ArrayList<>();
        for (int i = startVerse; i <= endVerse; i++) {
            output.add(verse(i));
        }
        return String.join("\n", output);
    }

    String sing() {
        return verses(1, 12);
    }
}

File: go/twelve-days/twelve_days.go
package twelve

import (
	"fmt"
	"strings"
)

const firstVerse = "On the %v day of Christmas my true love gave to me: "

var lyrics = []string{
	"twelve Drummers Drumming, ",
	"eleven Pipers Piping, ",
	"ten Lords-a-Leaping, ",
	"nine Ladies Dancing, ",
	"eight Maids-a-Milking, ",
	"seven Swans-a-Swimming, ",
	"six Geese-a-Laying, ",
	"five Gold Rings, ",
	"four Calling Birds, ",
	"three French Hens, ",
	"two Turtle Doves, ",
	"and ",
	"a Partridge in a Pear Tree.",
}

var days = map[int]string{
	1:  "first",
	2:  "second",
	3:  "third",
	4:  "fourth",
	5:  "fifth",
	6:  "sixth",
	7:  "seventh",
	8:  "eighth",
	9:  "ninth",
	10: "tenth",
	11: "eleventh",
	12: "twelfth",
}

func Verse(i int) string {
	out := []string{fmt.Sprintf(firstVerse, days[i])}
	start, end := len(lyrics)-i-1, len(lyrics)
	if i == 1 {
		start++
	}
	out = append(out, lyrics[start:end]...)
	return strings.Join(out, "")
}

func Song() string {
	out := []string{}
	for i := 1; i <= 12; i++ {
		out = append(out, Verse(i))
	}
	return strings.Join(out, "\n")
}
